File: pipeline/preprocessing.py
"""Photo -> rectified working-resolution map image.

Steps: raw (non-EXIF-rotated) load -> downscale -> paper-frame detection &
perspective rectification -> best-effort magnetic-north-line detection
(local scale anchor). Real photos, not scans: perspective, glare, shadows,
creases are expected, so every geometric step has a documented fallback
rather than raising on the happy-path assumption.
"""

from __future__ import annotations

from dataclasses import dataclass

import cv2
import numpy as np

from pipeline.config import WORKING_MAX_SIDE


@dataclass
class PreprocessResult:
    source_path: str
    image: np.ndarray                # rectified, working-resolution BGR image
    scale_to_original: float         # multiply working-res px by this to get original-photo px
    quad_found: bool                 # False => paper-frame detection fell back to full frame
    mn_line_xs: list[float]          # x positions (working-res px) of detected magnetic-north lines
    mn_line_spacing_px: float | None  # median spacing between them, None if <2 found


def load_image_exif_safe(path: str) -> np.ndarray:
    """Load a JPEG, return BGR uint8 ndarray.

    Deliberately does NOT apply the file's EXIF orientation tag. Checked
    directly against this project's test photos: the raw pixel data is
    already upright (title readable, map north-ish-up) while the EXIF
    orientation tags are inconsistent (6 on most files, 3 on one) and, when
    applied, rotate the *correctly-oriented* raw pixels into a wrong
    orientation -- i.e. the tags are stale/wrong for this dataset, not the
    pixels. Reading orientation doesn't matter to this pipeline anyway:
    find_paper_quad below establishes the working frame from the paper's own
    edges, not from metadata, so this is robust regardless of what a given
    photo's EXIF claims.

    Note: cv2.imread applies EXIF orientation automatically as of OpenCV 4.5+,
    which is exactly the behaviour we're opting out of here -- pass
    IMREAD_IGNORE_ORIENTATION explicitly to get genuinely raw pixels.
    """
    return cv2.imread(path, cv2.IMREAD_COLOR | cv2.IMREAD_IGNORE_ORIENTATION)


def white_balance(img: np.ndarray, percentile: float = 97.0) -> np.ndarray:
    """Normalize color cast using the photographed paper's own white margin
    as a reference (percentile white-patch balance): find the brightest
    pixels in the frame (almost certainly the white paper background, not
    map ink) and scale each channel so their average becomes neutral.

    Real photos of these maps carry a strong, inconsistent color cast from
    whatever indoor lighting they were shot under -- checked directly:
    map4.jpg's paper background averages BGR (185, 196, 217), ~30 points
    short of neutral on blue -- which pushes every warm map color (browns,
    tans) further toward "red" and collides with the course-overprint hue
    range in course_detection.py. A *fixed* gain correction wouldn't work
    here for the same reason fixed HSV thresholds don't (per-photo lighting
    varies too much) -- this instead measures and corrects each photo's own
    cast from its own paper.
    """
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    threshold = np.percentile(gray, percentile)
    white_mask = gray >= threshold
    if white_mask.sum() < 100:
        return img
    b, g, r = cv2.split(img.astype(np.float32))
    mb, mg, mr = b[white_mask].mean(), g[white_mask].mean(), r[white_mask].mean()
    target = (mb + mg + mr) / 3.0
    b *= target / max(mb, 1e-6)
    g *= target / max(mg, 1e-6)
    r *= target / max(mr, 1e-6)
    return np.clip(cv2.merge([b, g, r]), 0, 255).astype(np.uint8)


def downscale(img: np.ndarray, max_side: int = WORKING_MAX_SIDE) -> tuple[np.ndarray, float]:
    """Resize so the longer side is <= max_side. Returns (small_img, scale).

    scale = small_size / original_size (<= 1.0). To map a working-res
    coordinate back to the original photo: orig = working / scale.
    """
    h, w = img.shape[:2]
    longer = max(h, w)
    if longer <= max_side:
        return img, 1.0
    scale = max_side / longer
    small = cv2.resize(img, (int(round(w * scale)), int(round(h * scale))), interpolation=cv2.INTER_AREA)
    return small, scale


def _order_corners(pts: np.ndarray) -> np.ndarray:
    """Order 4 points as top-left, top-right, bottom-right, bottom-left."""
    pts = pts.reshape(4, 2).astype(np.float32)
    s = pts.sum(axis=1)
    diff = np.diff(pts, axis=1).reshape(-1)
    tl = pts[np.argmin(s)]
    br = pts[np.argmax(s)]
    tr = pts[np.argmin(diff)]
    bl = pts[np.argmax(diff)]
    return np.array([tl, tr, br, bl], dtype=np.float32)


def find_paper_quad(img: np.ndarray) -> np.ndarray | None:
    """Find the 4 corners of the photographed paper sheet.

    Largest-contour + polygon approximation against the table background.
    Returns ordered (tl, tr, br, bl) points in image pixel coords, or None if
    no confident quad was found (caller falls back to using the full frame).
    """
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    blurred = cv2.GaussianBlur(gray, (5, 5), 0)
    edges = cv2.Canny(blurred, 40, 120)
    edges = cv2.dilate(edges, np.ones((5, 5), np.uint8), iterations=2)

    contours, _ = cv2.findContours(edges, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    if not contours:
        return None
    largest = max(contours, key=cv2.contourArea)
    img_area = img.shape[0] * img.shape[1]
    if cv2.contourArea(largest) < 0.25 * img_area:
        # Sheet doesn't dominate the frame enough to trust this contour.
        return None

    # approxPolyDP on the *raw* contour, not its convex hull: verified
    # directly against map2.jpg/map4.jpg -- a shadow or low-contrast patch
    # near one corner puts a concave notch in the raw contour there, and
    # approxPolyDP happily returns a "clean" 4-gon that cuts straight across
    # the notch instead of reaching the true corner, silently discarding a
    # wedge of real map content (confirmed by cross-checking the found quad
    # against the actual photo). The hull bridges that kind of notch.
    hull = cv2.convexHull(largest)
    hull_area = cv2.contourArea(hull)
    peri = cv2.arcLength(hull, True)
    quad = None
    for eps_frac in (0.01, 0.02, 0.03, 0.05, 0.08):
        approx = cv2.approxPolyDP(hull, eps_frac * peri, True)
        if len(approx) == 4:
            quad = approx
            break

    # Even off the hull, approxPolyDP can still return a too-simple 4-gon
    # that undercuts the true boundary -- checked directly: this ratio was
    # ~0.69 on both map2.jpg and map4.jpg where a corner got cut off, vs.
    # >0.9 on files where the quad was fine. Below the threshold, prefer the
    # rotated bounding rectangle: it may include a sliver of table
    # background, but that's a far safer failure than silently cropping out
    # real map content.
    if quad is not None and cv2.contourArea(quad) >= 0.85 * hull_area:
        return _order_corners(quad)

    rect = cv2.minAreaRect(hull)
    box = cv2.boxPoints(rect)
    return _order_corners(box)


def rectify(img: np.ndarray, quad: np.ndarray) -> np.ndarray:
    """Perspective-warp `img` so `quad` maps onto an axis-aligned rectangle."""
    tl, tr, br, bl = quad
    width_top = np.linalg.norm(tr - tl)
    width_bottom = np.linalg.norm(br - bl)
    height_left = np.linalg.norm(bl - tl)
    height_right = np.linalg.norm(br - tr)
    out_w = int(round(max(width_top, width_bottom)))
    out_h = int(round(max(height_left, height_right)))
    out_w, out_h = max(out_w, 1), max(out_h, 1)

    dst = np.array([[0, 0], [out_w - 1, 0], [out_w - 1, out_h - 1], [0, out_h - 1]], dtype=np.float32)
    m = cv2.getPerspectiveTransform(quad, dst)
    return cv2.warpPerspective(img, m, (out_w, out_h))


def _all_segments(lines: np.ndarray | None, min_len: float) -> list[tuple[float, float, float, float]]:
    """Return (angle_deg_mod_180, x_mid, y_mid, length) for every long-enough
    HoughLinesP segment, with no angle restriction -- rectify() only aligns
    the *paper edges* to the frame, not the map's reading orientation, so the
    magnetic-north lines can come out horizontal just as easily as vertical
    depending on which corner the corner-ordering heuristic called "top-left".
    """
    out = []
    if lines is None:
        return out
    for seg in lines:
        # OpenCV has varied between returning (N, 1, 4) and (N, 4) across
        # versions/bindings for HoughLinesP; handle both.
        x1, y1, x2, y2 = np.asarray(seg).reshape(-1).astype(float)
        dx, dy = x2 - x1, y2 - y1
        length = float(np.hypot(dx, dy))
        if length < min_len:
            continue
        angle = float(np.degrees(np.arctan2(dy, dx)) % 180.0)  # 0=horizontal, 90=vertical
        out.append((angle, (x1 + x2) / 2.0, (y1 + y2) / 2.0, length))
    return out


def _drop_border_segments(
    segments: list[tuple[float, float, float, float]], img_shape: tuple[int, int], margin_frac: float = 0.03
) -> list[tuple[float, float, float, float]]:
    """Drop segments whose midpoint sits within `margin_frac` of any image
    edge. The rectified sheet's own outer border/frame otherwise dominates
    Hough results (it's the single longest, straightest edge in the photo)
    and would get mistaken for part of the magnetic-north-line family.
    """
    h, w = img_shape[:2]
    mx, my = w * margin_frac, h * margin_frac
    return [s for s in segments if mx <= s[1] <= w - mx and my <= s[2] <= h - my]


def _mask_edges(mask: np.ndarray) -> np.ndarray:
    """Thin a filled color mask down to boundary edges before Hough.

    Feeding a filled blob (e.g. a lake) straight into HoughLinesP makes every
    row of interior pixels look like a candidate line; running on the mask's
    boundary instead means a lake contributes its (curved, non-repeating)
    shoreline rather than a false family of parallel lines.
    """
    mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, np.ones((7, 7), np.uint8))
    return cv2.Canny(mask, 50, 150)


def detect_magnetic_north_lines(img: np.ndarray, tol_deg: float = 10.0) -> tuple[list[float], float | None]:
    """Best-effort detection of the equally-spaced magnetic-north lines.

    Color-agnostic (grayscale Canny -> Hough), restricted to near-vertical
    segments: find_paper_quad + rectify already leave the sheet close to
    axis-aligned, so this is a safer assumption than searching for a dominant
    line family at *any* angle -- that was tried and, on real photos full of
    other straight linework (building edges, hatching, paths), it just as
    often locks onto a spurious diagonal family as the real one. Getting this
    wrong would rotate an already-correct image into a wrong one, which is a
    worse outcome than simply not finding the lines.

    This is explicitly a best-effort component (see plan): the real photos
    interrupt these thin printed lines constantly (contours, paths, course
    overprint, text crossing them), so returning ([], None) is common and
    does not block the rest of the pipeline -- spacing is only ever used as
    an optional local-scale hint.

    Returns (sorted x-positions px, median spacing px | None if <2 lines).
    """
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    edges = cv2.Canny(gray, 30, 90)
    lines = cv2.HoughLinesP(edges, 1, np.pi / 360, threshold=50,
                             minLineLength=img.shape[0] * 0.06, maxLineGap=90)
    segs = _drop_border_segments(_all_segments(lines, min_len=img.shape[0] * 0.06), img.shape)

    near_vertical = [s for s in segs if abs(((s[0] - 90.0 + 90.0) % 180.0) - 90.0) <= tol_deg]
    if len(near_vertical) < 3:
        return [], None

    near_vertical.sort(key=lambda s: s[1])
    clusters: list[list[tuple[float, float, float, float]]] = [[near_vertical[0]]]
    for s in near_vertical[1:]:
        if s[1] - clusters[-1][-1][1] <= img.shape[1] * 0.01:
            clusters[-1].append(s)
        else:
            clusters.append([s])

    # A real magnetic-north line gets chopped into many short Hough segments
    # by everything crossing it (contours, paths, course overprint, text),
    # but those fragments collectively span most of the map's height. A
    # stray feature that merely happens to have a near-vertical edge (a
    # building corner, a lake shore) contributes only one or two segments
    # over a small y-range -- require both a minimum member count and a
    # minimum vertical span to tell the two apart.
    img_h = img.shape[0]
    centers = sorted(
        float(np.mean([s[1] for s in cluster]))
        for cluster in clusters
        if len(cluster) >= 3 and (max(s[2] for s in cluster) - min(s[2] for s in cluster)) >= 0.25 * img_h
    )
    if len(centers) < 3:
        return centers, None
    spacings = np.diff(centers)
    median_spacing = float(np.median(spacings))
    keep = [centers[0]]
    for x, sp in zip(centers[1:], spacings):
        if abs(sp - median_spacing) / max(median_spacing, 1e-6) < 0.4:
            keep.append(x)
    spacing_out = float(np.median(np.diff(keep))) if len(keep) >= 2 else None
    return keep, spacing_out


def preprocess_image(path: str, max_side: int = WORKING_MAX_SIDE) -> PreprocessResult:
    full = load_image_exif_safe(path)
    small, scale = downscale(full, max_side)

    quad = find_paper_quad(small)
    if quad is not None:
        rectified = rectify(small, quad)
        quad_found = True
    else:
        rectified = small
        quad_found = False

    rectified = white_balance(rectified)
    mn_xs, mn_spacing = detect_magnetic_north_lines(rectified)

    return PreprocessResult(
        source_path=path,
        image=rectified,
        scale_to_original=scale,
        quad_found=quad_found,
        mn_line_xs=mn_xs,
        mn_line_spacing_px=mn_spacing,
    )

File: app/rust/src/api/preprocessing.rs
//! Photo -> rectified, color-normalized image. Conceptual Rust port of the
//! Phase 0 Python prototype's `preprocessing.find_paper_quad` /
//! `preprocessing.rectify` / `preprocessing.white_balance` /
//! `preprocessing.detect_magnetic_north_lines` (see
//! `python_prototype/pipeline/preprocessing.py` and `PHASE0_HANDOFF.md`) --
//! not a line-for-line port. Reading-orientation correction is deliberately
//! NOT reimplemented here: the Python prototype's version is a hardcoded
//! per-file rotation table (`config.PAGE_ROTATION_K`) that `PHASE0_HANDOFF.md`
//! explicitly says not to carry into Phase 1 as-is, since it doesn't
//! generalize to a photo this repo hasn't seen -- that knob belongs in the
//! manual-correction UI instead.

use opencv::core::{self, Mat, Point, Point2f, Scalar, Size, Vec4i, Vector};
use opencv::prelude::*;
use opencv::{imgcodecs, imgproc};

const WORKING_MAX_SIDE: i32 = 3000;

#[derive(Debug, Clone)]
pub struct RectifyResult {
    /// Rectified (or, if no confident quad was found, merely downscaled)
    /// image, PNG-encoded.
    pub image_png: Vec<u8>,
    /// False => paper-quad detection fell back to using the full frame.
    pub quad_found: bool,
    pub width: i32,
    pub height: i32,
    /// X positions (px, in the output image's coordinates) of detected
    /// magnetic-north lines. Best-effort -- frequently empty, see
    /// `detect_magnetic_north_lines`.
    pub mn_line_xs: Vec<f32>,
    /// Median spacing (px) between `mn_line_xs`, if at least 2 were found.
    pub mn_line_spacing_px: Option<f32>,
    /// `working_resolution_size / original_photo_size` (<=1.0) -- multiply a
    /// working-resolution pixel coordinate by `1.0 / scale_to_original` to
    /// recover its position in the original, full-resolution photo.
    pub scale_to_original: f32,
}

pub fn rectify_photo(image_bytes: Vec<u8>) -> Result<RectifyResult, String> {
    rectify_photo_impl(image_bytes).map_err(|e| e.to_string())
}

/// Result of the full preprocessing chain, kept as an in-memory `Mat` (not
/// yet PNG-encoded) so downstream stages (`segmentation`, `course_detection`)
/// can run against the same rectified image without a decode round-trip.
/// `#[frb(ignore)]` because this is purely an internal handoff type between
/// Rust modules -- flutter_rust_bridge scans everything under `crate::api`
/// regardless of `pub`/`pub(crate)`, so without this it leaks into the Dart
/// API surface as an opaque class nobody should call.
#[flutter_rust_bridge::frb(ignore)]
pub(crate) struct Preprocessed {
    pub rectified: Mat,
    pub quad_found: bool,
    pub mn_line_xs: Vec<f32>,
    pub mn_line_spacing_px: Option<f32>,
    /// `working_resolution_size / original_photo_size` (<=1.0) -- multiply a
    /// working-resolution pixel coordinate by `1.0 / scale_to_original` to
    /// recover its position in the original, full-resolution photo. Matches
    /// the Python prototype's `PreprocessResult.scale_to_original`.
    pub scale_to_original: f32,
}

pub(crate) fn preprocess(image_bytes: &[u8]) -> opencv::Result<Preprocessed> {
    let buf = Vector::<u8>::from_slice(image_bytes);
    let img = imgcodecs::imdecode(
        &buf,
        imgcodecs::IMREAD_COLOR | imgcodecs::IMREAD_IGNORE_ORIENTATION,
    )?;
    if img.empty() {
        return Err(opencv::Error::new(
            core::StsError,
            "could not decode image bytes".to_string(),
        ));
    }

    let (small, scale_to_original) = downscale(&img, WORKING_MAX_SIDE)?;

    let quad = find_paper_quad(&small)?;
    let (rectified, quad_found) = match quad {
        Some(q) => (rectify(&small, &q)?, true),
        None => (small, false),
    };
    let rectified = white_balance(&rectified, 97.0)?;
    let (mn_line_xs, mn_line_spacing_px) = detect_magnetic_north_lines(&rectified, 10.0)?;

    Ok(Preprocessed {
        rectified,
        quad_found,
        mn_line_xs,
        mn_line_spacing_px,
        scale_to_original: scale_to_original as f32,
    })
}

fn rectify_photo_impl(image_bytes: Vec<u8>) -> opencv::Result<RectifyResult> {
    let pre = preprocess(&image_bytes)?;

    let mut out_buf = Vector::<u8>::new();
    imgcodecs::imencode_def(".png", &pre.rectified, &mut out_buf)?;

    Ok(RectifyResult {
        width: pre.rectified.cols(),
        height: pre.rectified.rows(),
        image_png: out_buf.to_vec(),
        quad_found: pre.quad_found,
        mn_line_xs: pre.mn_line_xs,
        mn_line_spacing_px: pre.mn_line_spacing_px,
        scale_to_original: pre.scale_to_original,
    })
}

/// Returns (downscaled image, scale) where `scale = downscaled_size /
/// original_size` (<=1.0) -- multiply a working-resolution pixel coordinate
/// by `1.0 / scale` to recover its position in the original, full-resolution
/// photo. Matches the Python prototype's `preprocessing.downscale`.
fn downscale(img: &Mat, max_side: i32) -> opencv::Result<(Mat, f64)> {
    let (w, h) = (img.cols(), img.rows());
    let longer = w.max(h);
    if longer <= max_side {
        return Ok((img.clone(), 1.0));
    }
    let scale = max_side as f64 / longer as f64;
    let mut small = Mat::default();
    imgproc::resize(
        img,
        &mut small,
        Size::new(
            (w as f64 * scale).round() as i32,
            (h as f64 * scale).round() as i32,
        ),
        0.0,
        0.0,
        imgproc::INTER_AREA,
    )?;
    Ok((small, scale))
}

/// Order 4 corner points as (top-left, top-right, bottom-right, bottom-left).
fn order_corners(pts: &[Point2f; 4]) -> [Point2f; 4] {
    let (mut tl, mut br, mut tr, mut bl) = (pts[0], pts[0], pts[0], pts[0]);
    let (mut min_s, mut max_s, mut min_d, mut max_d) = (f32::MAX, f32::MIN, f32::MAX, f32::MIN);
    for p in pts {
        let s = p.x + p.y;
        let d = p.y - p.x;
        if s < min_s {
            min_s = s;
            tl = *p;
        }
        if s > max_s {
            max_s = s;
            br = *p;
        }
        if d < min_d {
            min_d = d;
            tr = *p;
        }
        if d > max_d {
            max_d = d;
            bl = *p;
        }
    }
    [tl, tr, br, bl]
}

/// Largest-contour + convex-hull polygon approximation, with a
/// rotated-bounding-box fallback -- see `preprocessing.find_paper_quad`'s
/// docstring in the Python prototype for why the hull (not the raw contour)
/// is used, and why an undercut approximation falls back to `minAreaRect`
/// rather than being trusted as-is.
fn find_paper_quad(img: &Mat) -> opencv::Result<Option<[Point2f; 4]>> {
    let mut gray = Mat::default();
    imgproc::cvt_color_def(img, &mut gray, imgproc::COLOR_BGR2GRAY)?;
    let mut blurred = Mat::default();
    imgproc::gaussian_blur_def(&gray, &mut blurred, Size::new(5, 5), 0.0)?;
    let mut edges = Mat::default();
    imgproc::canny_def(&blurred, &mut edges, 40.0, 120.0)?;
    let kernel = Mat::new_rows_cols_with_default(5, 5, core::CV_8U, Scalar::all(1.0))?;
    let mut dilated = Mat::default();
    imgproc::dilate(
        &edges,
        &mut dilated,
        &kernel,
        Point::new(-1, -1),
        2,
        core::BORDER_CONSTANT,
        Scalar::all(f64::MAX),
    )?;

    let mut contours = Vector::<Vector<Point>>::new();
    imgproc::find_contours_def(
        &dilated,
        &mut contours,
        imgproc::RETR_EXTERNAL,
        imgproc::CHAIN_APPROX_SIMPLE,
    )?;
    if contours.is_empty() {
        return Ok(None);
    }

    let img_area = (img.rows() as f64) * (img.cols() as f64);
    let mut largest = contours.get(0)?;
    let mut largest_area = imgproc::contour_area_def(&largest)?;
    for c in contours.iter().skip(1) {
        let area = imgproc::contour_area_def(&c)?;
        if area > largest_area {
            largest_area = area;
            largest = c;
        }
    }
    if largest_area < 0.25 * img_area {
        return Ok(None);
    }

    let mut hull = Vector::<Point>::new();
    imgproc::convex_hull_def(&largest, &mut hull)?;
    let hull_area = imgproc::contour_area_def(&hull)?;
    let peri = imgproc::arc_length(&hull, true)?;

    let mut quad: Option<Vector<Point>> = None;
    for eps_frac in [0.01, 0.02, 0.03, 0.05, 0.08] {
        let mut approx = Vector::<Point>::new();
        imgproc::approx_poly_dp(&hull, &mut approx, eps_frac * peri, true)?;
        if approx.len() == 4 {
            quad = Some(approx);
            break;
        }
    }

    if let Some(q) = &quad {
        let quad_area = imgproc::contour_area_def(q)?;
        if quad_area >= 0.85 * hull_area {
            let pts: Vec<Point2f> = q.iter().map(|p| Point2f::new(p.x as f32, p.y as f32)).collect();
            return Ok(Some(order_corners(&[pts[0], pts[1], pts[2], pts[3]])));
        }
    }

    let rect = imgproc::min_area_rect(&hull)?;
    let mut box_mat = Mat::default();
    imgproc::box_points(rect, &mut box_mat)?;
    let box_mat = box_mat.reshape_def(Point2f::opencv_channels())?;
    let pts = [
        *box_mat.at::<Point2f>(0)?,
        *box_mat.at::<Point2f>(1)?,
        *box_mat.at::<Point2f>(2)?,
        *box_mat.at::<Point2f>(3)?,
    ];
    Ok(Some(order_corners(&pts)))
}

fn point_dist(a: Point2f, b: Point2f) -> f32 {
    ((a.x - b.x).powi(2) + (a.y - b.y).powi(2)).sqrt()
}

/// Perspective-warp `img` so `quad` (tl, tr, br, bl) maps onto an
/// axis-aligned rectangle.
fn rectify(img: &Mat, quad: &[Point2f; 4]) -> opencv::Result<Mat> {
    let [tl, tr, br, bl] = *quad;
    let width_top = point_dist(tr, tl);
    let width_bottom = point_dist(br, bl);
    let height_left = point_dist(bl, tl);
    let height_right = point_dist(br, tr);
    let out_w = width_top.max(width_bottom).round().max(1.0) as i32;
    let out_h = height_left.max(height_right).round().max(1.0) as i32;

    let src = Vector::<Point2f>::from_slice(&[tl, tr, br, bl]);
    let dst = Vector::<Point2f>::from_slice(&[
        Point2f::new(0.0, 0.0),
        Point2f::new((out_w - 1) as f32, 0.0),
        Point2f::new((out_w - 1) as f32, (out_h - 1) as f32),
        Point2f::new(0.0, (out_h - 1) as f32),
    ]);
    let src_mat = Mat::from_slice(src.as_slice())?;
    let dst_mat = Mat::from_slice(dst.as_slice())?;
    let m = imgproc::get_perspective_transform_def(&src_mat, &dst_mat)?;

    let mut warped = Mat::default();
    imgproc::warp_perspective_def(img, &mut warped, &m, Size::new(out_w, out_h))?;
    Ok(warped)
}

/// Normalize color cast using the photographed paper's own white margin as a
/// reference (percentile white-patch balance): find the brightest pixels in
/// the frame (almost certainly the white paper background, not map ink) and
/// scale each channel so their average becomes neutral. Port of
/// `preprocessing.white_balance` in the Python prototype -- see its
/// docstring for why a *fixed* gain correction wouldn't work here (per-photo
/// lighting cast varies too much; this measures and corrects each photo's
/// own cast from its own paper, not a global constant).
fn white_balance(img: &Mat, percentile: f64) -> opencv::Result<Mat> {
    let mut gray = Mat::default();
    imgproc::cvt_color_def(img, &mut gray, imgproc::COLOR_BGR2GRAY)?;
    let gray_bytes = gray.data_bytes()?;

    let mut hist = [0u32; 256];
    for &v in gray_bytes {
        hist[v as usize] += 1;
    }
    let total: u32 = hist.iter().sum();
    let target_count = (total as f64 * percentile / 100.0).ceil() as u32;
    let mut threshold = 255u8;
    let mut cum = 0u32;
    for (i, &c) in hist.iter().enumerate() {
        cum += c;
        if cum >= target_count {
            threshold = i as u8;
            break;
        }
    }

    let img_bytes = img.data_bytes()?;
    let (mut sum_b, mut sum_g, mut sum_r, mut count) = (0u64, 0u64, 0u64, 0u64);
    for (i, &g) in gray_bytes.iter().enumerate() {
        if g >= threshold {
            let px = i * 3;
            sum_b += img_bytes[px] as u64;
            sum_g += img_bytes[px + 1] as u64;
            sum_r += img_bytes[px + 2] as u64;
            count += 1;
        }
    }
    if count < 100 {
        return Ok(img.clone());
    }

    let (mb, mg, mr) = (
        sum_b as f64 / count as f64,
        sum_g as f64 / count as f64,
        sum_r as f64 / count as f64,
    );
    let target = (mb + mg + mr) / 3.0;
    let gains = [target / mb.max(1e-6), target / mg.max(1e-6), target / mr.max(1e-6)];

    let mut out = img.clone();
    let out_bytes = out.data_bytes_mut()?;
    for px in out_bytes.chunks_exact_mut(3) {
        for (c, gain) in gains.iter().enumerate() {
            px[c] = (px[c] as f64 * gain).round().clamp(0.0, 255.0) as u8;
        }
    }
    Ok(out)
}

#[derive(Debug, Clone, Copy)]
struct MnSegment {
    /// Degrees, folded into [0, 180): 0 = horizontal, 90 = vertical.
    angle_deg: f64,
    mid_x: f64,
    mid_y: f64,
}

/// Convert raw HoughLinesP output into (angle, midpoint) segments, dropping
/// anything shorter than `min_len`. No angle restriction here -- the caller
/// filters to near-vertical, and can only assume that's the right family to
/// look for because reading orientation is expected to already be roughly
/// fixed (paper edges axis-aligned) by the time this runs.
fn all_segments(lines: &Vector<Vec4i>, min_len: f64) -> Vec<MnSegment> {
    let mut out = Vec::new();
    for seg in lines.iter() {
        let (x1, y1, x2, y2) = (seg[0] as f64, seg[1] as f64, seg[2] as f64, seg[3] as f64);
        let (dx, dy) = (x2 - x1, y2 - y1);
        let length = (dx * dx + dy * dy).sqrt();
        if length < min_len {
            continue;
        }
        let angle_deg = dy.atan2(dx).to_degrees().rem_euclid(180.0);
        out.push(MnSegment { angle_deg, mid_x: (x1 + x2) / 2.0, mid_y: (y1 + y2) / 2.0 });
    }
    out
}

/// Drop segments whose midpoint sits within `margin_frac` of any image edge
/// -- the rectified sheet's own outer border is otherwise the single
/// longest, straightest edge in the photo and would get mistaken for part of
/// the magnetic-north-line family.
fn drop_border_segments(segments: Vec<MnSegment>, width: i32, height: i32, margin_frac: f64) -> Vec<MnSegment> {
    let (w, h) = (width as f64, height as f64);
    let (mx, my) = (w * margin_frac, h * margin_frac);
    segments
        .into_iter()
        .filter(|s| s.mid_x >= mx && s.mid_x <= w - mx && s.mid_y >= my && s.mid_y <= h - my)
        .collect()
}

fn median(values: &[f64]) -> f64 {
    let mut sorted = values.to_vec();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let n = sorted.len();
    if n == 0 {
        0.0
    } else if n % 2 == 1 {
        sorted[n / 2]
    } else {
        (sorted[n / 2 - 1] + sorted[n / 2]) / 2.0
    }
}

/// Best-effort detection of the equally-spaced magnetic-north lines. Port of
/// `preprocessing.detect_magnetic_north_lines` in the Python prototype --
/// see its docstring for why this deliberately only searches near-vertical
/// (rather than any-angle) line families, and why returning nothing is a
/// common, acceptable outcome rather than something to work around: real
/// photos interrupt these thin printed lines constantly (contours, paths,
/// course overprint, text crossing them), and nothing downstream depends on
/// this succeeding -- spacing is only ever used as an optional local-scale
/// hint.
///
/// Returns (sorted x-positions px, median spacing px | None if <2 lines).
fn detect_magnetic_north_lines(img: &Mat, tol_deg: f64) -> opencv::Result<(Vec<f32>, Option<f32>)> {
    let (width, height) = (img.cols(), img.rows());
    let mut gray = Mat::default();
    imgproc::cvt_color_def(img, &mut gray, imgproc::COLOR_BGR2GRAY)?;
    let mut edges = Mat::default();
    imgproc::canny_def(&gray, &mut edges, 30.0, 90.0)?;

    let min_len = height as f64 * 0.06;
    let mut lines = Vector::<Vec4i>::new();
    imgproc::hough_lines_p(
        &edges,
        &mut lines,
        1.0,
        std::f64::consts::PI / 360.0,
        50,
        min_len,
        90.0,
    )?;

    let segs = all_segments(&lines, min_len);
    let segs = drop_border_segments(segs, width, height, 0.03);

    let near_vertical: Vec<MnSegment> = segs.into_iter().filter(|s| (s.angle_deg - 90.0).abs() <= tol_deg).collect();
    if near_vertical.len() < 3 {
        return Ok((vec![], None));
    }

    let mut sorted = near_vertical;
    sorted.sort_by(|a, b| a.mid_x.partial_cmp(&b.mid_x).unwrap());

    let cluster_gap = width as f64 * 0.01;
    let mut clusters: Vec<Vec<MnSegment>> = vec![vec![sorted[0]]];
    for s in sorted.into_iter().skip(1) {
        let last = clusters.last().unwrap().last().unwrap();
        if s.mid_x - last.mid_x <= cluster_gap {
            clusters.last_mut().unwrap().push(s);
        } else {
            clusters.push(vec![s]);
        }
    }

    // A real magnetic-north line gets chopped into many short Hough segments
    // by everything crossing it, but those fragments collectively span most
    // of the map's height. A stray feature that merely happens to have a
    // near-vertical edge contributes only one or two segments over a small
    // y-range -- require both a minimum member count and a minimum vertical
    // span to tell the two apart.
    let img_h = height as f64;
    let mut centers: Vec<f64> = clusters
        .into_iter()
        .filter(|c| {
            if c.len() < 3 {
                return false;
            }
            let (mut min_y, mut max_y) = (f64::MAX, f64::MIN);
            for s in c {
                min_y = min_y.min(s.mid_y);
                max_y = max_y.max(s.mid_y);
            }
            (max_y - min_y) >= 0.25 * img_h
        })
        .map(|c| c.iter().map(|s| s.mid_x).sum::<f64>() / c.len() as f64)
        .collect();
    centers.sort_by(|a, b| a.partial_cmp(b).unwrap());

    if centers.len() < 3 {
        return Ok((centers.iter().map(|&x| x as f32).collect(), None));
    }

    let spacings: Vec<f64> = centers.windows(2).map(|w| w[1] - w[0]).collect();
    let median_spacing = median(&spacings);
    let mut keep = vec![centers[0]];
    for (x, sp) in centers[1..].iter().zip(spacings.iter()) {
        if (sp - median_spacing).abs() / median_spacing.max(1e-6) < 0.4 {
            keep.push(*x);
        }
    }
    let spacing_out = if keep.len() >= 2 {
        let keep_spacings: Vec<f64> = keep.windows(2).map(|w| w[1] - w[0]).collect();
        Some(median(&keep_spacings) as f32)
    } else {
        None
    };
    Ok((keep.into_iter().map(|x| x as f32).collect(), spacing_out))
}
